### PR TITLE
Optimize dynamic string interpolation for symbol/true/false/nil/0-9

### PR DESCRIPTION
**Pull request by jeremyevans**
This provides a significant speedup for symbol, true, false,
nil, and 0-9, and a small speedup in most other cases.

Speedups (using included benchmarks):
:symbol        :: 60%
0-9            :: 50%
nil/true/false :: 20%
integer        :: 10%
[]             :: 10%
""             :: 3%

One reason this approach is faster is it can significantly reduce
the number of VM instructions for each interpolated value:

Before:

```
0005 dup
0006 checktype                              T_STRING
0008 branchif                               14
0010 dup
0011 opt_send_without_block                 <calldata!mid:to_s, argc:0, FCALL|ARGS_SIMPLE>
0013 tostring
```

After:

```
0005 tostring                               <calldata!mid:to_s, argc:0, ARGS_SIMPLE>
```

Initial idea, approach, and benchmarks from Eric Wong. I applied
the same approach against the master branch, updating it to handle
the significant internal changes since this was first proposed 4
years ago (such as CALL_INFO/CALL_CACHE -> CALL_DATA). I also
expanded it to optimize true/false/nil/0-9, and added handling of
missing methods, refined methods, and RUBY_DEBUG.

This requires making a few functions non-static, and adding some
non-static functions. rb_vm_call0_body is added as a non-static
function to appease Visual Studio 2019, fixing this error:

```
../src/vm.c(4167): error C2129: static function 'VALUE vm_call0_body(rb_execution_context_t *,rb_calling_info *,const VALUE *)' declared but not defined
```

The YJIT support for optimizing the tostring instruction is
disabled by this until YJIT is updated for the new tostring
instruction. This also disables 4 YJIT tests to fix the following
failures due to disabling the tostring instruction optimization:

```
Expected no exits, but got
{:tostring=>1}
```

Implements [Feature #13715]

Co-authored-by: Eric Wong <e@80x24.org>

### CODE REVIEW

**Comment by ioquatix** [comment]
This is amazing, well done!

**Comment by krzysiek1507** [comment]
I wonder if such improvement can be backported to the previous versions?

File: compile.c [hunk 1]
```diff
@@ -7640,17 +7640,17 @@ compile_evstr(rb_iseq_t *iseq, LINK_ANCHOR *const ret, const NODE *const node, i
...
-	ADD_INSN(ret, line_node, tostring);
-	ADD_LABEL(ret, isstr);
+        ADD_INSN(ret, line_node, dup);
+
+        if (ISEQ_COMPILE_DATA(iseq)->option->specialized_instruction) {
+            ADD_INSN1(ret, line_node, opt_to_s_unescaping,
+                      new_callinfo(iseq, idTo_s, 0, VM_CALL_FCALL, NULL, FALSE));
+        }
+        else {
+            ADD_SEND(ret, line_node, idTo_s, INT2FIX(0));
```
[the remark this answers is not in the document]
Comment (ko1) [reply compile.c:7651]: Thank you @XrXr and @jeremyevans 
I realized this issue after my comment.

How about to use the following 2 instructions?

```
objtostring (call `to_s` if it is not `T_STRING`)
anytostring (force to make String, master's tostring)
```

Not sure `objtostring` is appropriate though...

[the remark this answers is not in the document]
Comment (jeremyevans) [reply compile.c]: I reverted the previous commit, then added a commit to rename `tostring` to `objtostring`, then rebased on the current master branch.  If CI passes, I'm going to commit this.

**Comment by jeremyevans** [comment]
CI is failing for typeprof (https://github.com/jeremyevans/ruby/runs/4241059870?check_suite_focus=true#step:15:1947): `Error: test: smoke/once.rb(TypeProf::SmokeTest): RuntimeError: Unknown insn: objtostring`

@mame Can you update typeprof to handle renaming the instruction to `objtostring`?

**Comment by mame** [comment]
@jeremyevans Sure, I will do it tomorrow morning.

**Pull request merged**